File: UnityScript/Example/Assets/OpenGUI/Scripts/OGDropDown.js
#pragma strict

import System.Collections.Generic;

class OGDropDown extends OGWidget {	
	// Classes
	public class DropDownItem {
		public var name : String;
		public var message : String;
		public var tickable : boolean = false;
		public var isTicked : boolean = false;
	}
	
	public class DropDownItemNested extends DropDownItem {
		public var tickOverrides : boolean = false;
	}
	
	public class DropDownItemRoot extends DropDownItem {
		public var nestedMenu : DropDownItemNested[];
	}
	
	// Vars
	public var title : String;
	public var target : GameObject;
	public var padding : RectOffset;
	public var submenu : DropDownItemRoot[];
	public var isDown : boolean = false;

	private var container : GameObject;
	private var background : OGSlicedSprite;
	private var label : OGLabel;
	private var activeNestedMenu : int = -1;
	private var timeStamp : float;
	
	
	////////////////////
	// Interaction
	////////////////////
	override function OnMouseUp () {
		CheckTicked ();

		isDown = !isDown;
		SetSelectedItem ( -1 );
		activeNestedMenu = -1;

		SetDrawn ( isDrawn );
	}

	override function OnMouseCancel () {
		Exit ();
	}

	// Exit
	private function Exit () {
		if ( isDown ) {
			isDown = false;
			SetSelectedItem ( -1 );

			SetDrawn ( isDrawn );
		}
	}

	// Check ticked
	private function CheckTicked () {
		var i : int = 0;
		var o : int = 0;
		var btn : OGListItem;

		for ( i = 0; i < container.transform.childCount; i++ ) {
			btn = container.transform.GetChild(i).Find("Button").GetComponent(OGListItem);
			btn.isTicked = submenu[i].isTicked;
			var nested : Transform = container.transform.GetChild(i).Find("NestedMenu");
			for ( o = 0; o < nested.childCount; o++ ) {
				btn = nested.GetChild(o).GetComponent(OGListItem);
				btn.isTicked = submenu[i].nestedMenu[o].isTicked;
			}
		}
	}

	// Set selected item
	private function SetSelectedItem ( i : int ) {
		var o : int = 0;
		
		for ( o = 0; o < container.transform.childCount; o++ ) {
			var entry : Transform = container.transform.GetChild(o);
			entry.Find("Button").GetComponent(OGListItem).selected = o == i;
			
			for ( var k = 0; k < entry.Find("NestedMenu").childCount; k++ ) {
				entry.Find("NestedMenu").GetChild(k).GetComponent(OGListItem).selected = false;
			}
		}
	}

	// Menu item
	public function SelectItem ( n : String ) {
		var i : int = int.Parse ( n );
		var item : DropDownItemRoot = submenu[i];
		
		if ( !String.IsNullOrEmpty ( item.message ) ) {
			target.SendMessage ( item.message );
		}
		
		if ( item.tickable ) {
			item.isTicked = !item.isTicked;
			container.transform.GetChild(i).Find("Button").GetComponent(OGListItem).isTicked = item.isTicked;
		}
	}

	public function HoverItem ( n : String ) {
		var i : int = int.Parse ( n );

		SetSelectedItem ( i );

		if ( submenu[i].nestedMenu.Length > 0 ) {
			activeNestedMenu = i;
		} else {
			activeNestedMenu = -1;
		}
		
		SetDrawn ( isDrawn );
	}

	// Nested item
	public function SelectNestedItem ( n : String ) {
		if ( activeNestedMenu < 0 ) { 
			Debug.LogWarning ( "OGDropDown | Nested menu out of bounds!" );
			return;
		}
		
		var i : int = int.Parse ( n );
		var item : DropDownItemNested = submenu[activeNestedMenu].nestedMenu[i];
		var nested : Transform = container.transform.GetChild(activeNestedMenu).Find("NestedMenu");

		if ( item.tickable ) {
			if ( item.tickOverrides ) {
				for ( var o : int = 0; o < submenu[activeNestedMenu].nestedMenu.Length; o++ ) {
					submenu[activeNestedMenu].nestedMenu[o].isTicked = o == i;
					nested.GetChild(o).GetComponent(OGListItem).isTicked = o == i;
				}
			} else {
				item.isTicked = !item.isTicked;
				nested.GetChild(i).GetComponent(OGListItem).isTicked = item.isTicked;
			}
		}

		if ( !String.IsNullOrEmpty ( item.message ) ) {
			target.SendMessage ( item.message );
		}
	}
	
	
	////////////////////
	// Set drawn
	////////////////////
	override function SetDrawn ( drawn : boolean ) {
		var i : int = 0;
		var o : int = 0;
		var lbl : OGLabel;
		
		isDrawn = drawn;
		
		label.isDrawn = isDrawn;	
		background.isDrawn = isDown && isDrawn;			
	
		// Submenu
		for ( i = 0; i < container.transform.childCount; i++ ) {
			var entry : Transform = container.transform.GetChild(i);
			var button : OGListItem = entry.Find("Button").GetComponent(OGListItem);	
			button.SetDrawn ( isDrawn && isDown );

			// Nested
			var nested : Transform = entry.Find("NestedMenu");
			var thisActive : boolean = i == activeNestedMenu;
			for ( o = 0; o < nested.childCount; o++ ) {
				var btn : OGListItem = nested.GetChild(o).GetComponent(OGListItem);
				btn.SetDrawn ( isDrawn && isDown && thisActive );	
			}

			var bg : OGSlicedSprite = entry.Find("Background").GetComponent(OGSlicedSprite);	
			bg.isDrawn = isDrawn && isDown && thisActive;
		}

		SetDirty ();
	}

	
	////////////////////
	// Build 
	////////////////////
	override function Build () {
		isSelectable = true;
		
		var i : int = 0;
		var o : int = 0;

		// Submenu container
		if ( !container && !FindChild("Submenu") ) {
			container = new GameObject ( "Submenu" );
			container.transform.parent = this.transform;
		
		} else if ( container == null ) {
			container = FindChild("Submenu");
		}

		container.transform.localPosition = new Vector3 ( 0, 1 + (padding.top*2)/this.transform.localScale.y, 0 );
		container.transform.localScale = Vector3.one;
		container.transform.localEulerAngles = Vector3.zero;
		
		// Submenu
		if ( submenu == null ) {
			submenu = new DropDownItemRoot[0];
		}
		
		// ^ Edit existing or create new ones
		for ( i = 0; i < submenu.Length; i++ ) {
			// Entry container
			var entry : Transform;

			if ( i < container.transform.childCount ) {
				entry = container.transform.GetChild(i);
			} else {
				entry = new GameObject ( i + ": " + submenu[i].name ).transform;
				entry.parent = container.transform;
			}

			entry.gameObject.name = i + ": " + submenu[i].name;
			entry.localPosition = new Vector3 ( 0, i, 0 );
			entry.localEulerAngles = Vector3.zero;
			entry.localScale = Vector3.one;

			// Entry button
			var button : OGListItem;

			if ( entry.Find("Button") ) {
				button = entry.Find("Button").GetComponent(OGListItem);
			} else {
				button = new GameObject ( "Button", OGListItem ).GetComponent(OGListItem);
				button.transform.parent = entry;
			}

			button.transform.localPosition = new Vector3 ( padding.left/this.transform.localScale.x, 0, 2 );
			button.transform.localScale = new Vector3 ( 1-(padding.left+padding.right)/this.transform.localScale.x, 1, 1 );
			button.transform.localEulerAngles = Vector3.zero;
			
			button.text = submenu[i].name;
			button.target = this.gameObject;
			button.message = "SelectItem";
			button.hoverMessage = "HoverItem";
			button.argument = i.ToString();
			button.hidden = true;
			button.styles.basic = this.styles.active;
			button.styles.active = this.styles.hover;
			button.styles.ticked = this.styles.ticked;

			// Nested menu container
			var nested : Transform;

			if ( entry.Find("NestedMenu") ) {
				nested = entry.Find("NestedMenu").transform;
			} else {
				nested = new GameObject ( "NestedMenu" ).transform;
				nested.parent = entry;
			}

			nested.localPosition = new Vector3 ( 1 - (padding.right*2)/this.transform.localScale.x, 0, 0 );
			nested.localScale = Vector3.one;
			nested.localEulerAngles = Vector3.zero;

			// Submenu
			// ^ Edit existing or create new ones
			for ( o = 0; o < submenu[i].nestedMenu.Length; o++ ) {
				var btn : OGListItem;
				var nestedItem : DropDownItemNested = submenu[i].nestedMenu[o];

				if ( o < nested.childCount ) {
					btn = nested.GetChild(o).GetComponent(OGListItem);
				} else {
					btn = new GameObject ( "", OGListItem ).GetComponent(OGListItem);
					btn.transform.parent = nested;
				}

				btn.gameObject.name = o + ": " + nestedItem.name;
				btn.text = nestedItem.name;
				btn.target = this.gameObject;
				btn.message = "SelectNestedItem";
				btn.argument = o.ToString();
				btn.hidden = true;
				btn.styles.basic = this.styles.active;
				btn.styles.active = this.styles.hover;
				btn.styles.ticked = this.styles.ticked;

				btn.transform.localPosition = new Vector3 ( padding.left/this.transform.localScale.x, o, 0 );
				btn.transform.localScale = new Vector3 ( 1-(padding.left+padding.right)/this.transform.localScale.x, 1, 1 );
				btn.transform.localEulerAngles = Vector3.zero;
			}

			// ^ Destroy remaining
			if ( nested.childCount > submenu[i].nestedMenu.Length ) {
				for ( o = submenu[i].nestedMenu.Length; o < nested.childCount; o++ ) {
					DestroyImmediate ( nested.GetChild(o).gameObject );
				}
			}

			// Nested background
			var bg : OGSlicedSprite;

			if ( entry.Find("Background") ) {
				bg = entry.Find("Background").GetComponent(OGSlicedSprite);
			} else {
				bg = new GameObject ( "Background", OGSlicedSprite ).GetComponent(OGSlicedSprite);
				bg.transform.parent = entry;
			}

			bg.styles.basic = this.styles.basic;
			bg.transform.localPosition = new Vector3 ( 1 - (padding.right*2)/this.transform.localScale.x, -padding.top/this.transform.localScale.y, 1 );
			if ( nested.childCount < 1 ) {
				bg.transform.localScale = Vector3.one;
			} else {
				bg.transform.localScale = new Vector3 ( 1, nested.childCount + (padding.top+padding.bottom)/this.transform.localScale.y, 1 );
			}
			bg.transform.localEulerAngles = Vector3.zero;
			bg.hidden = true;
		}

		// ^ Destroy remaining
		if ( container.transform.childCount > submenu.Length ) {
			for ( i = submenu.Length; i < container.transform.childCount; i++ ) {
				DestroyImmediate ( container.transform.GetChild(i).gameObject );
			}
		}

		// Background
		if ( background == null ) {
			if ( this.transform.Find( "SlicedSprite" ) ) {
				background = this.transform.Find( "SlicedSprite" ).GetComponent(OGSlicedSprite);
				
			} else {			
				var newSprite : OGSlicedSprite = new GameObject ( "SlicedSprite", OGSlicedSprite ).GetComponent ( OGSlicedSprite );
				newSprite.transform.parent = this.transform;
				background = newSprite;
			}
		}
			
		
		background.hidden = true;
		background.styles.basic = this.styles.basic;
		background.transform.localPosition = new Vector3 ( 0, 1 + padding.top/this.transform.localScale.y, 3 );
		background.transform.localEulerAngles = Vector3.zero;
		background.transform.localScale = new Vector3 ( 1, container.transform.childCount + (padding.top+padding.bottom)/this.transform.localScale.y, 1 );
		
		// Label
		if ( label == null ) {
			if ( this.gameObject.GetComponentInChildren ( OGLabel ) ) {
				label = this.gameObject.GetComponentInChildren ( OGLabel );
				
			} else {				
				var newLabel : OGLabel = new GameObject ( "Label", OGLabel ).GetComponent ( OGLabel );
				newLabel.transform.parent = this.transform;
				label = newLabel;
			}
		}

		label.text = title;
		label.styles.basic = this.styles.basic;		
		
		label.transform.localScale = new Vector3 ( ( label.lineWidth + label.styles.basic.text.padding.left + label.styles.basic.text.padding.right + 5 ) / this.transform.lossyScale.x, 1, 1 );
		label.transform.localEulerAngles = Vector3.zero;
		label.transform.localPosition = Vector3.zero;
		
		label.hidden = true;

		// Set drawn
		SetDrawn ( isDrawn );
	}


	////////////////////
	// Update
	////////////////////
	override function UpdateWidget () {
		// Null check
		if ( !background || !label || !container ) {
			Build ();
			return;
		}

		// Mouse
		mouseRct = label.drawRct;
	}
}

File: OpenGUI/Assets/Plugins/OpenGUI/Scripts/OGDropDown.cs
using UnityEngine;
using System.Collections;

using System.Collections.Generic;

public class OGDropDown : OGWidget {	
	// Classes
	[System.Serializable]
	public class DropDownItem {
		public string name;
		public string message;
		public string argument;
		public bool tickable = false;
		public bool isTicked = false;
	}
	
	[System.Serializable]
	public class DropDownItemNested : DropDownItem {
		public bool tickoverrides = false;
	}
	
	[System.Serializable]
	public class DropDownItemRoot : DropDownItem {
		public DropDownItemNested[] nestedMenu;
	}
	
	// Vars
	public string title = "";
	public GameObject target = null;
	public float nestedOffset = -10;
	public DropDownItemRoot[] submenu = null;
	public bool isDown = false;

	private int activeNestedMenu = -1;
	private float timeStamp;
	
	
	////////////////////
	// Interaction
	////////////////////
	override public void OnMouseUp () {
		if ( isDown ) {
			// Is the mouse over any nested menu items?
			if ( activeNestedMenu != -1 ) {
				for ( int n = 0; n < submenu[activeNestedMenu].nestedMenu.Length; n++ ) {
					if ( CheckMouseOver ( GetNestedItemRect ( n ) ) ) {
						SelectNestedItem ( n );
						return;
					}
				}
			}
			
			// Is the mouse over any root menu items?
			for ( int s = 0; s < submenu.Length; s++ ) {
				if ( CheckMouseOver ( GetRootItemRect ( s ) ) ) {
					SelectRootItem ( s );
					return;
				}
			}

			// Nope, exit
			Exit ();

		} else {
			isDown = true;
		
		}
	}

	override public void OnMouseCancel () {
		Exit ();
	}

	// Exit
	private void Exit () {
		isDown = false;
		activeNestedMenu = -1;

		root.ReleaseWidget ();
	}

	// Menu item
	public void SelectRootItem ( int i ) {
		DropDownItemRoot item = submenu[i];
		
		if ( !string.IsNullOrEmpty ( item.message ) ) {
			if ( !string.IsNullOrEmpty ( item.argument ) ) {
				target.SendMessage ( item.message, item.argument );
			} else {
				target.SendMessage ( item.message );
			}
		}
		
		if ( item.tickable ) {
			item.isTicked = !item.isTicked;
			activeNestedMenu = -1;
			Exit ();

		} else if ( item.nestedMenu.Length > 0 ) {
			if ( activeNestedMenu == i ) {
				activeNestedMenu = -1;
			} else {
				activeNestedMenu = i;
			}

		} else {
			Exit ();

		}
		
	}

	// Nested item
	public void SelectNestedItem ( int i ) {
		if ( activeNestedMenu < 0 ) { 
			Debug.LogWarning ( "OGDropDown | Nested menu out of bounds!" );
			return;
		}
		
		DropDownItemNested item = submenu[activeNestedMenu].nestedMenu[i];

		if ( item.tickable ) {
			if ( item.tickoverrides ) {
				for ( int o = 0; o < submenu[activeNestedMenu].nestedMenu.Length; o++ ) {
					submenu[activeNestedMenu].nestedMenu[o].isTicked = o == i;
				}
			} else {
				item.isTicked = !item.isTicked;
			}
		}
		
		Exit ();

		if ( !string.IsNullOrEmpty ( item.message ) ) {
			if ( !string.IsNullOrEmpty ( item.argument ) ) {
				target.SendMessage ( item.message, item.argument );
			} else {
				target.SendMessage ( item.message );
			}
		}
	}


	////////////////////
	// Rects
	////////////////////
	private Rect GetMouseRect () {
		return new Rect ( drawRct.x, drawRct.y, drawRct.width / 2, drawRct.height );
	}
	
	private Rect GetRootBackgroundRect () {
		return new Rect ( drawRct.x, drawRct.y - submenu.Length * drawRct.height - styles.active.text.padding.top - styles.active.text.padding.bottom, drawRct.width, drawRct.height * submenu.Length + styles.active.text.padding.top + styles.active.text.padding.bottom ); 
	}

	private Rect GetNestedBackgroundRect () {
		return new Rect ( drawRct.x + drawRct.width + nestedOffset, drawRct.y - ( activeNestedMenu + submenu[activeNestedMenu].nestedMenu.Length ) * drawRct.height - styles.active.text.padding.top - styles.active.text.padding.bottom, drawRct.width, drawRct.height * submenu[activeNestedMenu].nestedMenu.Length + styles.active.text.padding.bottom + styles.active.text.padding.top );
	}

	private Rect GetRootItemRect ( int i ) {
		return new Rect ( drawRct.x, drawRct.y - ( ( 1 + i ) * drawRct.height ) - styles.active.text.padding.top, drawRct.width, drawRct.height );
	}

	private Rect GetNestedItemRect ( int i ) {
		return new Rect ( drawRct.x + drawRct.width + nestedOffset, drawRct.y - ( 1 + activeNestedMenu + i ) * drawRct.height - styles.active.text.padding.top, drawRct.width, drawRct.height );
	}

	private Rect GetTickRect ( int i, bool isRoot ) {
		if ( isRoot ) {
			return new Rect ( drawRct.x + drawRct.width - drawRct.height - styles.ticked.text.padding.right, drawRct.y - ( ( 1 + i ) * drawRct.height + styles.ticked.text.padding.top ), drawRct.height, drawRct.height );
		} else {
			return new Rect ( drawRct.x + ( drawRct.width * 2 ) - drawRct.height - styles.ticked.text.padding.right + nestedOffset, drawRct.y - ( ( 1 + activeNestedMenu + i ) * drawRct.height + styles.ticked.text.padding.top ), drawRct.height, drawRct.height );
		}
	}
	

	////////////////////
	// Style
	////////////////////
	private OGStyle GetRootItemStyle ( int i ) {
		if ( CheckMouseOver ( GetRootItemRect ( i ) ) ) {
			return styles.hover;
		} else {
			return styles.active;
		}
	}

	private OGStyle GetNestedItemStyle ( int i ) {
		if ( CheckMouseOver ( GetNestedItemRect ( i ) ) ) {
			return styles.hover;
		} else {
			return styles.active;
		}
	}


	////////////////////
	// Update
	////////////////////
	override public void UpdateWidget () {
		// Persistent vars
		isSelectable = true;

		// Mouse
		if ( isDown ) {
			if ( activeNestedMenu != -1 ) {
				mouseRct = CombineRects ( GetRootBackgroundRect(), GetNestedBackgroundRect() );
			} else {
				mouseRct = GetRootBackgroundRect ();
			}
		} else {
			mouseRct = GetMouseRect();
		}

	}
	
	
	////////////////////
	// Draw
	////////////////////
	override public void DrawSkin () {
		if ( isDown ) {
			OGDrawHelper.DrawSlicedSprite ( GetRootBackgroundRect (), styles.active, drawDepth, tint, clipTo );
		
			// Draw item graphics
			for ( int s = 0; s < submenu.Length; s++ ) {
				if ( submenu[s].isTicked ) {
					OGDrawHelper.DrawSprite ( GetTickRect ( s, true ), styles.ticked, drawDepth, tint, clipTo );
				
				} else if ( submenu[s].nestedMenu.Length > 0 ) {
					OGDrawHelper.DrawSprite ( GetTickRect ( s, true ), styles.thumb, drawDepth, tint, clipTo );

				}

				if ( GetRootItemStyle ( s ) == styles.hover ) {
					OGDrawHelper.DrawSprite ( GetRootItemRect ( s ), styles.hover, drawDepth, tint, clipTo );
				}
			}

			if ( activeNestedMenu != -1 ) {
				OGDrawHelper.DrawSlicedSprite ( GetNestedBackgroundRect (), styles.active, drawDepth, tint, clipTo );
				
				for ( int n = 0; n < submenu[activeNestedMenu].nestedMenu.Length; n++ ) {
					if ( submenu[activeNestedMenu].nestedMenu[n].isTicked ) {
						OGDrawHelper.DrawSprite ( GetTickRect ( n, false ), styles.ticked, drawDepth, tint, clipTo );
					}
					
					if ( GetNestedItemStyle ( n ) == styles.hover ) {
						OGDrawHelper.DrawSprite ( GetNestedItemRect ( n ), styles.hover, drawDepth, tint, clipTo );
					}
				}
			}
		
		}
	}	

	override public void DrawText () {
		OGDrawHelper.DrawLabel ( drawRct, title, styles.basic.text, drawDepth, tint );

		if ( isDown ) {
			for ( int s = 0; s < submenu.Length; s++ ) {
				OGDrawHelper.DrawLabel ( GetRootItemRect ( s ), submenu[s].name, GetRootItemStyle ( s ).text, drawDepth, tint, clipTo );
			}

			if ( activeNestedMenu != -1 ) {
				for ( int n = 0; n < submenu[activeNestedMenu].nestedMenu.Length; n++ ) {
					OGDrawHelper.DrawLabel ( GetNestedItemRect ( n ), submenu[activeNestedMenu].nestedMenu[n].name, GetNestedItemStyle ( n ).text, drawDepth, tint, clipTo );
				}
			}
		}
	}
}
